src/cargo/core/compiler/mod.rs

```diff
@@ -1135,10 +1176,43 @@ fn on_stderr_line(
         struct CompilerMessage {
             rendered: String,
         }
-        if let Ok(error) = serde_json::from_str::<CompilerMessage>(compiler_message.get()) {
-            state.stderr(&error.rendered);
+        if let Ok(mut error) = serde_json::from_str::<CompilerMessage>(compiler_message.get()) {
+            // state.stderr will add a newline
+            if error.rendered.ends_with('\n') {
+                error.rendered.pop();
+            }
+            let rendered = if color {
+                error.rendered
+            } else {
+                strip_ansi_escapes::strip(&error.rendered)
+                    .map(|v| String::from_utf8(v).expect("utf8"))
+                    .unwrap_or(error.rendered)
+            };
+            state.stderr(rendered);
             return Ok(());
         }
+    } else {
+        // Remove color information from the rendered string. rustc has not
+        // included color in the past, so to avoid breaking anything, strip it
+        // out when --json-rendered=termcolor is used. This runs
+        // unconditionally under the assumption that Cargo will eventually
+        // move to this as the default mode. Perhaps in the future, cargo
+        // could allow the user to enable/disable color (such as with a
+        // `--json-rendered` or `--color` or `--message-format` flag).
```

Comment:
Ok so just to confirm, this is here because Cargo may unconditionally inject `--json-rendered=termcolor`, but the user isn't expecting that, so we have to reset output to what the user expects?